http.go

```
@@ -0,0 +1,101 @@
+// RFC 3161 3.4
+const TimestampQuery = "application/timestamp-query"
+
+// TimestampReply is the content-type of timestamp reply
+// RFC 3161 3.4
+const TimestampReply = "application/timestamp-reply"
+
+// httpTimestamper is a HTTP-based timestamper.
+type httpTimestamper struct {
+	rt       http.RoundTripper
+	endpoint string
+}
+
+// NewHTTPTimestamper creates a HTTP-based timestamper with the endpoint
+// provided by the TSA.
+// http.DefaultTransport is used if nil RoundTripper is passed.
+func NewHTTPTimestamper(rt http.RoundTripper, endpoint string) (Timestamper, error) {
+	if rt == nil {
+		rt = http.DefaultTransport
+	}
+	if _, err := url.Parse(endpoint); err != nil {
+		return nil, err
+	}
+	return &httpTimestamper{
+		rt:       rt,
+		endpoint: endpoint,
+	}, nil
+}
+
+// Timestamp sends the request to the remote TSA server for timestamping.
+// Reference: RFC 3161 3.4 Time-Stamp Protocol via HTTP
+func (ts *httpTimestamper) Timestamp(ctx context.Context, req *Request) (*Response, error) {
+	// prepare for http request
+	reqBytes, err := req.MarshalBinary()
+	if err != nil {
+		return nil, err
+	}
+	hReq, err := http.NewRequestWithContext(ctx, http.MethodPost, ts.endpoint, bytes.NewReader(reqBytes))
+	if err != nil {
+		return nil, err
+	}
+	hReq.Header.Set("Content-Type", TimestampQuery)
+
+	// send the request to the remote TSA server
+	hResp, err := ts.rt.RoundTrip(hReq)
+	if err != nil {
+		return nil, err
+	}
+	defer hResp.Body.Close()
+
+	// verify HTTP response
+	if hResp.StatusCode != http.StatusOK {
+		return nil, fmt.Errorf("%s %q: https response bad status: %s", http.MethodPost, ts.endpoint, hResp.Status)
+	}
+	if contentType := hResp.Header.Get("Content-Type"); contentType != TimestampReply {
+		return nil, fmt.Errorf("%s %q: unexpected response content type: %s", http.MethodPost, ts.endpoint, contentType)
+	}
+
+	// read response
+	body := io.LimitReader(hResp.Body, maxBodyLength)
+	respBytes, err := io.ReadAll(body)
```

If the response length exceeds `maxBodyLength`, we will not encounter an error here; instead, the error will occur during parsing. Would this make it difficult for users to debug? This is because the user might not know why the response is incomplete, and we don't inform the user of the actual reason in the error message.

---

We may need to fail fast here by checking the size of `respBytes`, which is equal to `maxBodyLength`

---

Updated.